src/Generator/TypeGenerator.php

```diff
@@ -149,18 +173,18 @@ private function __construct(array $types, bool $nullable)
      *
      * @psalm-return non-empty-string
      */
-    public function generate()
+    public function generate(): string
     {
         $typesAsStrings = array_map(
-            fn(AtomicType $type): string => $type->fullyQualifiedName(),
+            static fn (AtomicType $type): string => $type->fullyQualifiedName(),
             $this->types
         );
 
         if ($this->nullable) {
-            return '?' . implode('|', $typesAsStrings);
+            return '?' . implode($this->atomicTypesSeparator(), $typesAsStrings);
```

Comment:
Intersection types are not nullable as of PHP 8.1, right? Wonder if that should be validated somewhere? Or did I miss something somewhere?

Reply:
https://3v4l.org/pO7JhB#v8.1rc2

Reply:
Yeah, they are not nullable, but this entire `TypeGenerator` is quite lax about inputs, as its only job is really to just produce a new type.

In fact, union types are also quite wonky: https://3v4l.org/tkNtB#v8.0.11

The `TypeGenerator` still works with `?int|float` though, so the flaw is pre-existing